File: Subtree of Another Tree/Leetcode_572.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, S, T):
        return not (S or T) or (
            (S and T)
            and S.val == T.val
            and self.isSameTree(S.left, T.left)
            and self.isSameTree(S.right, T.right)
        )

    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        return s is not None and (
            self.isSameTree(s, t)
            or self.isSubtree(s.left, t)
            or self.isSubtree(s.right, t)
        )

File: Subtree of Another Tree/Leetcode_572.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
 
class Solution {
 public:
  bool isSameTree(TreeNode* s, TreeNode* t) {
    return (s == nullptr && t == nullptr) |
           (s != nullptr && t != nullptr && s->val == t->val &&
            isSameTree(s->left, t->left) && isSameTree(s->right, t->right));
  }

  bool isSubtree(TreeNode* s, TreeNode* t) {
    return s != nullptr &&
           (isSameTree(s, t) | isSubtree(s->left, t) | isSubtree(s->right, t));
  }
};
